overlord/devicestate/devicestate.go

```diff
@@ -380,11 +393,55 @@ func Remodel(st *state.State, new *asserts.Model) ([]*state.TaskSet, error) {
 			if err != nil {
 				return nil, err
 			}
-			addTss(ts)
+			tss = append(tss, ts)
 		} else if err != nil {
 			return nil, err
 		}
 	}
+	// TODO: Validate that all bases and default-providers are part
+	//       of the install tasksets and error if not. If the
+	//       prereq task handler check starts adding installs into
+	//       our remodel change our carefully constructed wait chain
+	//       breaks down.
+
+	// Ensure all download/check tasks are run *before* the install
+	// tasks. During a remodel the network may not be available so
+	// we need to ensure we have everything local.
+	var lastDownloadInChain, firstInstallInChain *state.Task
+	var prevDownload, prevInstall *state.Task
+	for _, ts := range tss {
+		// make sure all things happen sequentially:
+		// - all tasks inside ts{Download,Install} already wait for
+		//   each other
+		//   Out chains look like this:
+		//     install1 <- verify1 <- download1
+		//     install2 <- verify2 <- download2
+		// - add wait of each first ts{Download,Install} task for
+		//   the last previous ts{Download,Install} task
+		//   Our chains now looks like:
+		//     download1 (no waits)
+		//     download2 <- verify1
+		//     install1 <- verify1 <- download1
+		//     install2 <- verify2 <- install1
+		firstDownload, lastDownload, firstInstall, lastInstall := extractDownloadInstallEdgesFromTs(ts)
```

Comment:
these variable names are a bit confusing,  either

firstOfDowloand    or downloadStart    and   lastOfDownload   or downloadLast   etc